personhood/contract_spawner.go

```diff
@@ -0,0 +1,241 @@
+func (c *ContractSpawner) Spawn(rst byzcoin.ReadOnlyStateTrie, inst byzcoin.Instruction, coins []byzcoin.Coin) (sc []byzcoin.StateChange, cout []byzcoin.Coin, err error) {
+	cout = coins
+
+	var darcID darc.ID
+	_, _, _, darcID, err = rst.GetValues(inst.InstanceID.Slice())
+	if err != nil {
+		return
+	}
+
+	// Spawn creates a new coin account as a separate instance.
+	ca := inst.DeriveID("")
+	var instBuf []byte
+	cID := inst.Spawn.ContractID
+	log.Lvlf3("Spawning %s instance to %x", cID, ca.Slice())
+	switch cID {
+	case ContractSpawnerID:
+		c.parseArgs(inst.Spawn.Args)
+		instBuf, err = protobuf.Encode(&c.SpawnerStruct)
+		if err != nil {
+			return nil, nil, errors.New("couldn't encode SpawnerInstance: " + err.Error())
+		}
+	case byzcoin.ContractDarcID:
+		if err = c.getCoins(cout, c.CostDarc); err != nil {
+			return
+		}
+		instBuf = inst.Spawn.Args.Search("darc")
+		d, err := darc.NewFromProtobuf(instBuf)
+		if err != nil {
+			return nil, nil, err
+		}
+		ca = byzcoin.NewInstanceID(d.GetBaseID())
+	case contracts.ContractCoinID:
+		if err = c.getCoins(cout, c.CostCoin); err != nil {
+			return
+		}
+		coin := &byzcoin.Coin{
+			Name: byzcoin.NewInstanceID(inst.Spawn.Args.Search("coinName")),
+		}
+		for i := range cout {
+			if cout[i].Name.Equal(coin.Name) {
+				err = coin.SafeAdd(cout[i].Value)
+				if err != nil {
+					return nil, nil, err
+				}
+				log.Lvl2("Adding initial balance:", coin.Value)
+				err = cout[i].SafeSub(coin.Value)
+				if err != nil {
+					return nil, nil, err
+				}
+			}
+		}
+		darcID = inst.Spawn.Args.Search("darcID")
+		h := sha256.New()
+		h.Write([]byte("coin"))
+		h.Write(darcID)
+		ca = byzcoin.NewInstanceID(h.Sum(nil))
+		instBuf, err = protobuf.Encode(coin)
+		if err != nil {
+			return nil, nil, err
+		}
+	case ContractCredentialID:
+		if err = c.getCoins(cout, c.CostCredential); err != nil {
+			return
+		}
+		instBuf = inst.Spawn.Args.Search("credential")
+		var cred CredentialStruct
+		err = protobuf.Decode(instBuf, &cred)
+		if err != nil {
+			return nil, nil, err
+		}
+		darcID = inst.Spawn.Args.Search("darcID")
+		h := sha256.New()
+		h.Write([]byte("credential"))
+		h.Write(darcID)
+		ca = byzcoin.NewInstanceID(h.Sum(nil))
+	case ContractPopPartyID:
+		if err = c.getCoins(cout, c.CostParty); err != nil {
+			return
+		}
+		return ContractPopParty{}.Spawn(rst, inst, cout)
+	case ContractRoPaSciID:
+		if err = c.getCoins(cout, c.CostRoPaSci); err != nil {
+			return
+		}
+		return ContractRoPaSci{}.Spawn(rst, inst, cout)
+	default:
+		return nil, nil, errors.New("don't know how to spawn this type of contract")
+	}
+	sc = []byzcoin.StateChange{
+		byzcoin.NewStateChange(byzcoin.Create, ca, cID, instBuf, darcID),
+	}
+	return
+}
+
+func (c ContractSpawner) getCoins(coins []byzcoin.Coin, cost byzcoin.Coin) error {
+	if cost.Value == 0 {
+		return nil
+	}
+	for i := range coins {
+		if coins[i].Name.Equal(cost.Name) {
+			if coins[i].Value >= cost.Value {
+				coins[i].SafeSub(cost.Value)
```

Comment:
please don't ignore the err SafeSub returns. I see why you can get away with it here, but it's dangerous for a future reader/editor.